core/src/main/scala/com/salesforce/op/evaluators/OpMultiClassificationEvaluator.scala

```diff
@@ -133,20 +168,176 @@ private[op] class OpMultiClassificationEvaluator
+        case Row(label: Double, pred: Double, prob: DenseVector) => (label, pred, prob.toArray)
+      }
+      val confusionMatrixByThreshold = calculateConfMatrixMetricsByThreshold(rddCm)
+      val misClassifications = calculateMisClassificationMetrics( rddCm.map{ case (label, pred, _) => (label, pred)} )
+
       val metrics = MultiClassificationMetrics(
         Precision = precision,
         Recall = recall,
         F1 = f1,
         Error = error,
         ThresholdMetrics = thresholdMetrics,
-        TopKMetrics = topKMetrics
+        TopKMetrics = topKMetrics,
+        ConfusionMatrixMetrics = confusionMatrixByThreshold,
+        MisClassificationMetrics = misClassifications
       )
 
       log.info("Evaluated metrics: {}", metrics.toString)
       metrics
     }
   }
 
+/**
+ * function to construct the confusion matrix for the top n most occurring labels
+ * @param labelPredictionCtRDD RDD of ((label, prediction, confidence), count)
+ * @param cmClasses the top n most occurring classes, sorted by counts descending
+ * @return an org.apache.spark.mllib.linalg.Matrix class, representing the confusion matrix
+ */
+  def constructConfusionMatrix(
+    labelPredictionCtRDD: RDD[((Double, Double, Double), Long)],
+    cmClasses: Seq[Double]): Seq[Long] = {
+
+    val confusionMatrixMap = labelPredictionCtRDD.map {
+      case ((label, prediction, _), count) => ((label, prediction), count)
+    }.reduceByKey(_ + _)
+      .map {
+        case ((label, prediction), count) => (label, prediction) -> count
+      }.collectAsMap()
+
+    cmClasses.map(label => {
+      cmClasses.map(prediction => {
+        confusionMatrixMap.get((label, prediction)).getOrElse(0L)
+      })
+    }).flatten
+  }
+
+  object SearchHelper extends Serializable{
+
+    def findThreshold(arr: IndexedSeq[Double], element: Double): Double = {
+      if (element > arr.last) arr.last
+      else if (element < arr(0)) 0.0
+      else {
+        val insertionPoint = new SearchImpl(arr).search(element).insertionPoint
+        val insertionPointValue = arr(insertionPoint)
+        if (element == insertionPointValue) insertionPointValue
+        else arr(insertionPoint-1)
+      }
+    }
+  }
+
+/**
+ * function to calculate confusion matrix for TopK most occurring labels by confidence threshold
+ *
+ * @param data RDD of (label, prediction, prediction probability vector)
+ * @return sequence of Matrix that corresponds to the confMatrixThresholds
+*/
+
+  def calculateConfMatrixMetricsByThreshold(
+    data: RDD[(Double, Double, Array[Double])]): MulticlassConfMatrixMetricsByThreshold = {
+
+    val labelCountsRDD = data.map { case (label, _, _) => (label, 1L) }.reduceByKey(_ + _)
+    val cmClasses = labelCountsRDD.sortBy(-_._2).map(_._1).take($(confMatrixNumClasses)).toSeq
+    val cmClassesSet = new HashSet() ++ cmClasses
+
+    val dataTopNLabels = data.filter { case (label, _, _) => cmClassesSet.contains(label) }
```

Comment:
Is this equivalent to?
```
    val dataTopNLabels = data.filter { case (label, prediction, _) => 
      cmClassesSet.contains(label) || cmClassesSet.contains(prediction)
    }
```

Reply:
It equals to &&. I like this simpler syntax - will change it